### ISSUE TITLE
Sending pdu.Body

### ISSUE THREAD

**Issue by xintron** [fiorix/go-smpp#6]
From my understanding it's not yet possible to write anything over the wire if you use a receiver och transceiver which makes things difficult. Having the possibility to a HandlerFunction is great, but it's not that helpful if you have no way of replying back to the incoming PDU's.

One example being a transceiver where you use `tx.Submit(...)` which is great how it's blocking and taking care of responses and response matching for you, but when the DeliverSM finally arrives in the handler there is no way to reply to this with an DeliverSMResp (non that I've found, please correct me if I'm wrong).

My suggestion is to make all connections (Transmitter, Receiver and Transceiver) implement the Writer interface, that way the clients can respond to DeliverSM, send their own DeliverSM (acting as a server) and have their own EnquireLink handling in the future (right now EnquireLink only seems to work if the client initiates it; but the client will never reply to an enquire link from the server).

**Comment by fiorix** [comment]
That's correct. Although IMO we shouldn't let users of this package write their own PDUs over the wire to avoid breaking the connection due to malformed messages. Instead, we should implement methods such as `DeliverSMResp` and have the `HandlerFunc` pass the `Receiver` as the first argument before the `pdu.Body`.

Does that make sense to you?

**Comment by xintron** [comment]
I see the reason for wanting it that way. My concern is that it will make the library less flexible.

First of, you would need to send in the complete bind struct (tr, rx or trx) since in a rx scenario you would need it to reply to incoming DeliverSM. For tr it's the same but for incoming  SubmitSM.

The current tx.Submit() is good and abstracts away the complexity of waiting for the response and handles the matching of seq id. But I personally would want it to be non-blocking (handling the response myself). The main reason for this is to log and sync the SubmitSM being sent and then update/log again when I get the SubmitSMResp back. Best of all would be to have pre-made methods handling the asynchronous nature of the protocol and at the same time allow for free PDU writing and reading.

Based on own experience there are SMPP serverns not strictly following the protocol and I've had to extend the libraries I've used to handle those issues. If a method is made for everything this would make it impossible to solve; thus I put my vote on allowing the user to access writes manually.

The handler function should probably get two arguments. First one a context that wraps whatever underlying bind in use (tr, rx, trx) so that it's generic enough to use and the second the PDU. Thoughts on this?

**Comment by fiorix** [comment]
Got it. In `net/http` they provide a `Hijack` method that allow users to take the underlying `Conn` object and do whatever they want with the socket. In our case that would be the `Writer` you're looking for.

However, I still think we should have a bottom layer handling persistent connection for clients, as it is today, and enquire link messages. Same should be true for servers wrt enquire link.

As of now we already have a decent set of predefined PDUs in the `pdu/types.go` file and any of them can be extended by retrieving the body by calling `Fields`. We might need guidelines for using the `pdufield.Map` type so users don't get too lost.

**Comment by xintron** [comment]
That sound like a good solution.

Regarding standard events like enquirelink and automatically responding to those, as well as sending them. I think that would fit nicely as a sort of middleware. There can be pre-defined ones (like EnquirelinkHandler) that you chain together, putting your own handler last (the one every client will most likely need to implement). This way it would be possible to make DeliverySM responding an optional, automatic setup for those who wants that.

With easy write access on the connection it will also be possible to handle the Bind/Unbind and reconnect handling manually if needed. It should not be the preferred way nut allowing it would be great.

Extending the PDUs is great, so I think that should fit most people's needs.

**Comment by fiorix** [comment]
Cool, I have a much clear view on this now. For the server side we might want to mimic the `net/http` semantics with a `ServeMux` that let you assign handlers to arbitrary messages by calling `Handle` - done that in my go-diameter package already. Some messages, like enquire link, would have a default handler that is replaceable in case the user wants it. Same would apply for any other message such as bind/unbind and so on.

**Comment by xintron** [comment]
Should make it easy to make the library fit each users need. From those who just want to do a simple SubmitSM and not worry about the enquirelink and deliversm etc to this who want full control and handle every PDU in their own way

Should the connections be made more similar perhaps? The all should implement writing and reading over the wire so that the mux sent to all the handlers have an SMPPConn that implements the network related parts the same way (the different binds behave very similar so this would hopefully simplify things in the long run).

**Comment by fiorix** [comment]
Yes, we should expose the underlying `smpp.Conn` type that is already defined in `conn.go` to handlers.

**Comment by fiorix** [comment]
Closing as out of scope. Adding a Writer interface would require a significant API redesign.
